File: task_manager_lib/src/process_tree.rs
use std::collections::HashMap;
use crate::process::Process;
use serde::{Serialize, Deserialize};

#[derive(Serialize, Deserialize)]
pub struct ProcessTree {
    root: Process,
    children: HashMap<i32, Box<ProcessTree>>, // Store children using `pid` as key
}

impl ProcessTree {
    pub fn new(root: Process) -> Self {
        ProcessTree {
            root,
            children: HashMap::new(),
        }
    }

    pub fn get_root(&self) -> &Process {
        &self.root
    }

    pub fn add_child(&mut self, parent: Process, child: Process) -> Result<(), &'static str> {
        if self.root == parent {
            self.children
                .entry(child.get_pid())
                .or_insert(Box::new(ProcessTree::new(child)));
            Ok(())
        } else {
            for subtree in self.children.values_mut() {
                if subtree.add_child(parent.clone(), child.clone()).is_ok() {
                    return Ok(());
                }
            }
            Err("Parent not found")
        }
    }

    pub fn serialize(&self, filename: &str) -> Result<(), std::io::Error> {
        let file = std::fs::File::create(filename)?;
        serde_json::to_writer(file, self)?;
        Ok(())
    }

    pub fn exists(&self, pid: i32) -> bool {
        if self.root.get_pid() == pid {
            return true;
        }
        for subtree in self.children.values() {
            if subtree.exists(pid) {
                return true;
            }
        }
        false
    }
}

File: ProcessTree.cs
using System.Collections.Generic;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace WinTaskManager;
public class ProcessTree
{
    [JsonPropertyName("root")]
    public Process Root { get; set; }

    [JsonPropertyName("children")]
    public Dictionary<int, ProcessTree> Children { get; set; } = new Dictionary<int, ProcessTree>();

    public bool Exists(int pid)
    {
        if (Root.Pid == pid)
            return true;

        foreach (var subtree in Children.Values)
        {
            if (subtree.Exists(pid))
                return true;
        }

        return false;
    }

    public void AddChild(Process parent, Process child)
    {
        if (Root.Pid == parent.Pid)
        {
            Children[child.Pid] = new ProcessTree { Root = child };
        }
        else
        {
            foreach (var subtree in Children.Values)
            {
                if (subtree.Exists(parent.Pid))
                {
                    subtree.AddChild(parent, child);
                    return;
                }
            }
        }
    }
    public static ProcessTree Deserialize(string json)
    {
        return JsonSerializer.Deserialize<ProcessTree>(json);
    }
    public ProcessTree Find(int pid)
    {
        if (Root.Pid == pid)
            return this;

        foreach (var subtree in Children.Values)
        {
            var result = subtree.Find(pid);
            if (result != null)
                return result;
        }

        return null;
    }

    public string Serialize()
    {
        return JsonSerializer.Serialize(this, new JsonSerializerOptions { WriteIndented = true });
    }
}